frontend/src/components/worksheets/WorksheetItemList.js

```diff
@@ -187,6 +106,117 @@ class WorksheetItemList extends React.Component {
         }
     };
 
+    // Create a worksheet item based on props and add it to worksheet_items.
+    // - item: information about the table to display
+    // - index: integer representing the index in the list of items
+    // - focused: whether this item has the focus
+    // - editPermission: whether we have permission to edit this item
+    // - setFocus: call back to select this item
+    // - updateWorksheetSubFocusIndex: call back to notify parent of which row is selected (for tables)
+    addWorksheetItems = (props, worksheet_items, prevItem, afterItem) => {
+        var item = props.item;
+
+        // Determine URL corresponding to item.
+        var url = null;
+        if (
+            item.bundles_spec &&
+            item.bundles_spec.bundle_infos[0] &&
+            item.bundles_spec.bundle_infos[0].uuid
+        )
+            url = '/bundles/' + item.bundles_spec.bundle_infos[0].uuid;
+        if (item.subworksheet_info) url = '/worksheets/' + item.subworksheet_info.uuid;
+
+        props.key = props.id = 'codalab-worksheet-item-' + props.focusIndex;
+        props.url = url;
+        props.prevItem = prevItem;
+        props.after_sort_key = getAfterSortKey(
+            props.item,
+            props.item.mode === 'markup_block' ? undefined : props.subFocusIndex,
+        );
+        props.ids = getIds(item);
+        // showNewButtonsAfterEachBundleRow is set to true when we have a bundle table, because in this case,
+        // we must show the new upload / new run buttons after each row in the table (in the BundleRow component)
+        // as opposed to at the end of the table (in ItemWrapper).
+        props.showNewButtonsAfterEachBundleRow =
+            props.item.mode === 'table_block' && !props.item.loadedFromPlaceholder;
+        const constructor = BLOCK_TO_COMPONENT[item.mode];
+
+        // for a placeholder block, if the directive in current version is already being fetched
+        // no need to fetch the same directive again
+        let skipDirectiveFetch = false;
+        if (item.mode === 'placeholder_block') {
+            const directive = item.directive;
+            if (
+                props.version <= this.state.fetchingDirectives.length - 1 &&
+                this.state.fetchingDirectives[props.version].has(directive)
+            ) {
+                skipDirectiveFetch = true;
```

Comment:
It just feels like this caching logic is pretty complicated. Could we instead always call `fetchData`, but then in the implementation of `fetchData` itself, we add the caching logic? That way, we can keep the logic within react components much more simple.

Reply:
We could use something like `useSWR` (https://swr.vercel.app/getting-started) instead.